Framework/Source/API/D3D/D3D12/LowLevel/D3D12DescriptorSet.cpp

```diff
@@ -46,36 +46,43 @@ namespace Falcor
     bool DescriptorSet::apiInit()
     {
         mpApiData = std::make_shared<DescriptorSetApiData>();
+        uint32_t count = 0;
+        const auto& type = falcorToDxDescType(mLayout.getRange(0).type);
 
         // For each range we need to allocate a table from a heap
-        mpApiData->allocations.resize(mLayout.getRangeCount());
-        for (size_t i = 0; i < mpApiData->allocations.size(); i++)
+        mpApiData->rangeBaseOffset.resize(mLayout.getRangeCount());
+
+        for (size_t i = 0; i < mLayout.getRangeCount(); i++)
         {
             const auto& range = mLayout.getRange(i);
-            D3D12DescriptorHeap* pHeap = getHeap(mpPool.get(), range.type);
-            mpApiData->allocations[i] = pHeap->allocateDescriptors(range.descCount);
-            if (mpApiData->allocations[i] == false)
-            {
-                // Execute deferred releases and try again
-                mpPool->executeDeferredReleases();
-                mpApiData->allocations[i] = pHeap->allocateDescriptors(range.descCount);
-                if (!mpApiData->allocations[i])
-                {
-                    return false;
-                }
-            }
+            mpApiData->rangeBaseOffset[i] = count;
+            assert(type == falcorToDxDescType(range.type)); // We can only allocate from a single heap
+            count += range.descCount;
         }
-        return true;
+
+        const auto& range0 = mLayout.getRange(0);
+        D3D12DescriptorHeap* pHeap = getHeap(mpPool.get(), range0.type);
```

Comment:
It looks like `range0.type` here is already what you've got in the `type` variable above.

This helps me understand `DescriptorSetApiData` above a bit. It seems like each descriptor set gets restricted to a single type of descriptor, but can hold multiple ranges. It is a bit weird that the data type doesn't enforce that (e.g., why do we have to query `mLayout.getRange(0).type` and then `assert` that the others match, instead of having this be enforced so that there is just `mLayout.type` and all the ranges implicitly use that type?).

The `rangeBaseOffset` then seems to be where each of the ranges starts in the contiguous allocate that is backing the whole set. Simple enough.

Reply:
It's not the same type. One if the falcor type, the other is the D3D type.

This limitation is specific to D3D12. In Vulkan we can have different types. Note that I check that the API types match, not the Falcor types.